lib/panoptes/client/authentication.rb

```diff
@@ -0,0 +1,69 @@
+require 'deprecate'
+
+module Panoptes
+  class Client
+    module Authentication
+      extend Gem::Deprecate
+
+      def jwt_payload
+        raise NotLoggedIn unless @auth[:token]
+        payload, = JWT.decode @auth[:token], jwt_signing_public_key, algorithm: 'RS512'
+        payload
+      rescue JWT::ExpiredSignature
+        raise AuthenticationExpired
+      end
+
+      def token_contents
+        jwt_payload.fetch('data')
```

Comment:
My previous understanding of the expiry was incorrect, once a token's expiry is expired the underlying gem will raise an error. 

If I configure a valid token to use with the client that has 1.5 hours remaining before expiry then decoding the contents will be the same each time. As it stands every time we request a field from the contents payload we have to pay the cost to decode it. 

Is there any way to avoid the computational cost of decoding / verifying token integrity on each attribute access?

Reply:
We can definitely store the expiry date as promised in the JWT and compare to that instead of decoding the token every time, but then we don't *really* know if the token is still valid or if someone changed the value in our running process. I thought this approach was more secure but it is definitely more costly.

Reply:
We could also add a strict mode setting to force that the token always be checked (like this) or if it can read cached values.

Reply:
(but then the strict mode setting could still be fudged)

Reply:
> but then we don't really know if the token is still valid or if someone changed the value in our running process

The API for using this client is that a token is passed to the initializer, from then on it's read only for that client instance. My understanding is a user can't update the stored token for the client without creating a new client. 

I don't think we should be worried if an user/attacker changes an instance variable in a running process, if they do that's broken the public API of the client and so be it. 

> We can definitely store the expiry date as promised in the JWT and compare to that instead of decoding the token every time

I think checking the expiration validity when accessing the `token_contents` is a nice way to implement this. 

E.g. when decoding the token, store the expiry time for future use [like caesar does](https://github.com/zooniverse/caesar/blob/3d1585604c8efc404d3a58131e9cc23cfe5c26cb/app/models/credential.rb#L88). Then on each `token_contents` we can guard against expired tokens via an expiry time vs the current time check (ensuring UTC or local time zone comparison). If expired the memoized token variable is cleared and any future call to access token contents will have to decode it again and thus raise an expired token error. 

OOI we could benchmark both approaches quickly using https://ruby-doc.org/stdlib-2.4.5/libdoc/benchmark/rdoc/Benchmark.html to get a sense of the implementation costs vs performance. If decoding on access approach is about the same time time and it's simpler code than i'm all for it, my hunch is it'll be much slower. 
OOI we could benchmark both approaches quickly using https://ruby-doc.org/stdlib-2.4.5/libdoc/benchmark/rdoc/Benchmark.html to get a sense of the implementation costs vs performance. If decoding is about the sa

Reply:
yeah i think i was being paranoid. ultimately if they have an expired token they're not going to have any luck actually talking panoptes, whether or not they trick the client into sending the request.